### ISSUE TITLE
怎么插入到resnet中， 

### ISSUE THREAD

**Issue by jiafengshen** [koshian2/OctConv-TFKeras#1]
他的输入high 和 low 部分是怎么划分的

**Comment by koshian2** [comment]
@jiafengshen 
I do not understand Chinese well, but I think I'm asking about the structure of ResNet with OctConv, so I will explain.

First, let's look back on the operations of **OctConv**. This is from the paper, and we need high and low inputs and outputs respectively.

![octconv_02](https://user-images.githubusercontent.com/12977641/56612349-17d81f00-664f-11e9-9809-0be8187176cf.png)

And, this is a traditional ResNet structure. Let's call it **ResBlock**.

![ResBlock](https://user-images.githubusercontent.com/12977641/56612929-836ebc00-6650-11e9-95ff-88ec7d781e07.png)

**ResBlock with OctConv** is as follows. Just replace Conv with OctConv and make it parallel.

![2019-04-24 (2)](https://user-images.githubusercontent.com/12977641/56613919-adc17900-6652-11e9-925e-d9c92f1045fd.png)

How to make high and low in the first ResBlock, we apply average pooling to the input of the model to make it low. High uses the input of the model as it is.

![2019-04-24 (3)](https://user-images.githubusercontent.com/12977641/56614387-e1e96980-6653-11e9-8d14-6e91d05f6a03.png)

And in the last ResBlock, concat high and low, apply global average pooling, and pass it to the output layer activation function such as softmax.

![2019-04-24 (4)](https://user-images.githubusercontent.com/12977641/56614686-8c618c80-6654-11e9-8e04-6972e3adcbc8.png)

It is possible to apply global average pooling each of high and low followed by concat, but it is less successible (It is observed that the error does not go down and stagnates, or the accuracy suddenly drops during training) .

**Comment by AakashKumarNain** [comment]
@koshian2 excellent explanation. I have one question though. If I am correct, the `high` and `low` parts are chosen from the channel dimension with `1-alpha` and `alpha` respectively. If I am correct in the above visualization, you are taking `alpha==0`. Is that a correct assumption of mine?

**Comment by koshian2** [comment]
@AakashKumarNain
In the above figure, **ResBlock with OctConv** is a general case where alpha is 0 to 1.

If alpha is 0 or 1, it will be equal to the traditional ResBlock. Because if alpha is 0, it can be considered that there is no low sequence. Also, if alpha is 1, it can be considered that there is no high sequence.

To understand what OctConv is doing, the figure written in the paper is useful.
![](https://user-images.githubusercontent.com/12977641/56812416-cb225d00-6875-11e9-9897-b7e5b491b568.png)
In the traditional convolution (e.g. Conv2D), the kernel has a shape of  ```(k, k, c_in, c_out)```(This shape order changes depending on the framework). OctConv just assigns the Conv2D kernel to high or low depending on  alpha value. That is, if alpha is 0 or 1, 3 out of 4 cubes in OctConv's kernel will be meaningless (size = 0). Under these conditions, OctConv's kernel looks the same as Conv2D's kernel.

In other words, the traditional convolution can also interpret as a special condition (alpha = 0) in OctConv. Thus, **ResBlock with OctConv** is a general case where alpha is 0 to 1, right?

**Comment by AdityaSoni19031997** [comment]
Thanks for the explanation!

**Comment by AakashKumarNain** [comment]
@koshian2 Thanks for the explanation. That is much clear now. Also, can I use your implementation as a reference implementation for my work?

**Comment by koshian2** [comment]
@AakashKumarNain 
Sorry, my reply was late.
This project is a MIT license. You can use freely within it.

**Comment by koshian2** [comment]
I changed the model of the last layer because I slightly misunderstood (issue #7)
The last layer in upper figure I wrote is different now.
